Deriva support

hategan opened the pull request:
This, together with the corresponding dms PR (https://github.com/whole-tale/girder_wt_data_manager/pull/51), adds:
- support for importing data from remote zip/bdbags using io_over_http
- integration with deriva under api/v1/integrations/deriva
- DMS downloads from http(s)://&lt;loc>/&lt;path>.zip?path=<path_in_zip>, again, using io_over_http
- adds the deriva scope to Globus auth, which is needed to fetch files from deriva locations requiring authentication

I don't have a dashboard running, and I don't really understand the integrations thing (e.g., at what point in the integrations flow does the import actually happen?), so I tried to copy what I saw in other cases. That said, this can be tested by invoking the familiar importData endpoint with a link to a deriva bdbag, such as the one appearing in https://identifiers.fair-research.org/hdl:20.500.12633/11RHwdYqWNBZL, creating a session with the resulting folder, and mounting that session with  girderfs/wt_dms_fs. That's somewhat tedious, but I don't know of a better solution.

Xarthisius commented on server/lib/__init__.py:
```diff
@@ -24,19 +25,22 @@
 
 IMPORT_PROVIDERS = ImportProviders()
 IMPORT_PROVIDERS.addProvider(BDBagProvider())
+IMPORT_PROVIDERS.addProvider(DerivaProvider())
```
Order here matters unfortunately. `BDBagProvider` just checks if url/pid ends with `.zip`, which is true for Deriva resolved urls, e.g.:

```
https://identifiers.fair-research.org/hdl:20.500.12633/Uzn848Mmg1Wk ->
https://pbcconsortium.s3.amazonaws.com/wholetale/7c066384317e1395b28e082c66209a25/2021-12-01_13.38.24/Dataset_1-882P.zip
```

For now, I'd suggest pushing DerivaProvider before BDBagProvider until we figure out better solution.

hategan replied:
Yes, indeed! Will fix.

hategan replied:
Should be fixed. Not sure why github is not marking this as outdated.

Xarthisius commented on server/lib/bdbag/bdbag_provider.py:
```diff
@@ -54,23 +65,40 @@ def _add(self, path: pathlib.Path, url: Optional[str], orig_path: pathlib.Path,
-        with zipfile.ZipFile(pid) as zf:
-            with tempfile.TemporaryDirectory() as tmp_dir:
-                zf.extractall(tmp_dir)
+        if base_url:
+            pid = base_url + '/' + pid
```
`base_url` is a bait. This thing exists solely for DataONE integration and should never be used by anything else.
```suggestion
```

hategan replied:
Well, yes and no. The pattern is there in multiple places. Deriva has multiple deployments. So does/did the Globus thing.

Xarthisius replied:
That's ok. All I say is that `base_url` is used throughout our only client as something DataONE related, so it won't have any DERIVA specific values ever. AFAICT `pid` in this context is already a full blown url.

hategan replied:
It probably won't matter. The reason I'd keep it because, even though it's not used in practice, it makes sense in theory.

Xarthisius commented on server/lib/deriva/provider.py:
```diff
@@ -0,0 +1,26 @@
+        sz = -1
+        if 'size' in entity:
+            sz = entity['size']
+        name = None
```
Can we default to something nice like:

```
name = pathlib.Path(urlparse(entity.getValue()).path).name  # where urlparse is from urllib.parse
```

potentially with:

```
if name.endswith(".zip"):
    name = name[:-4]
```

hategan replied:
We probably should.

hategan commented on server/lib/bdbag/bdbag_provider.py:
```diff
@@ -12,9 +15,17 @@
 
+def _text(o: object) -> str:
+    if isinstance(o, str):
+        return str
```
How on earth?!

hategan commented on server/lib/bdbag/bdbag_provider.py:
```diff
@@ -54,23 +65,40 @@ def _add(self, path: pathlib.Path, url: Optional[str], orig_path: pathlib.Path,
 
     def matches(self, entity: Entity) -> bool:
-        return False
+        try:
+            return str(entity.getValue()).endswith('.zip')
+        except:
+            return False
```
Yeah, I don't see it either.

hategan commented on server/lib/bdbag/bdbag_provider.py:
```diff
@@ -89,42 +117,55 @@ def _listRecursive(self, user: Dict[str, object], pid: str, name: str,
-                    bag_file_path = bag_path / k
+                    bag_file_path: Path = bag_path.joinpath(k)
+                    zip_path = self._path_in_zip(bag_file_path)
+                    size = bag_file_path.root.getinfo(str(zip_path)).file_size  # type: ignore
+
+                    # alternatively, we should maybe import the zip file and then refer to
+                    # the file inside the zip using the same mechanism as for http(s) zips
+
+                    extracted = None
+                    if zip_url.startswith('file://'):
+                        extracted = bag_file_path.root.extract(str(zip_path), path=tmp_dir)
+                        url = 'file://' + extracted
+                    else:
+                        url = zip_url + '?path=' + str(zip_path)
+
                     yield ImportItem(ImportItem.FILE, k,
-                                     size=bag_file_path.stat().st_size,
+                                     size=size,
                                      mimeType='application/octet-stream',
-                                     url='file://{}'.format(bag_file_path.absolute()))
+                                     url=url)
 
-    def _get_zip_file(self, entity: Entity) -> zipfile.ZipFile:
-        if not isinstance(entity.getValue(), zipfile.ZipFile):
-            raise ValueError('Entity value must be a ZipFile')
-        return cast(zipfile.ZipFile, entity.getValue())
+                    if extracted:
+                        pathlib.Path(extracted).unlink()
 
-    def _read_fetch_txt(self, root: _BagTree, main: pathlib.Path) -> None:
+    def _read_fetch_txt(self, root: _BagTree, main: zipfile.Path) -> None:
         fetch_path = main / 'fetch.txt'
         if fetch_path.exists():
             with fetch_path.open() as f:
-                line = f.readline()
+                line = _text(f.readline())
                 while line:
                     self._parse_fetch_line(root, line.strip())
-                    line = f.readline()
+                    line = _text(f.readline())
```
Not all versions of zipfile support that, at least as far as I can tell. In particular, the one I have, which appears to be from Python 3.8, complains when zipfile.Path.open has anything else than 'r' or 'w'.

craig-willis commented on server/lib/deriva/provider.py:
```diff
@@ -0,0 +1,27 @@
+        super().__init__('DERIVA')
+
+    def matches(self, entity: Entity) -> bool:
+        return str(entity.getValue()).startswith('https://pbcconsortium.s3.amazonaws.com/')
```
Consider putting this in a configuration setting instead of hardcoding ala https://github.com/whole-tale/girder_wholetale/blob/master/server/lib/dataverse/provider.py#L96?  This way we can add new DERIVA installations without re-deploying.

craig-willis reviewed (changes requested):
I tested the following URLs:
*  https://girder.local.wholetale.org/api/v1/integration/deriva?url=https%3A%2F%2Fpbcconsortium.s3.amazonaws.com%2Fwholetale%2F5ad7cdf55b0d5007601015b7ff1ea8d6%2F2021-11-09_21.47.58%2FDataset_1-882P.zip&force=false
* https://girder.local.wholetale.org/api/v1/integration/deriva?url=https%3A%2F%2Fidentifiers.fair-research.org%2Fhdl%3A20.500.12633%2FUzn848Mmg1Wk

The basic import works, but I noticed the following:
* The dataset `identifier` is set to `unknown` on registration/import. I expected this to be set to the identifier of the bag (minid or URL?).   A side effect of this is if I download the bag and import it directly via `/importBag`, the dataset record is updated (provider changes from `DERIVA` to `BDBag`).  I expected to have two different datasets. I suggest adding a valid identifier value and checking it during import.
* A user without permission on the DERIVA dataset is able to import the bag, but reading a remote file (i.e., listed in fetch.txt) fails due to permission denied. (With fuse, this comes across as input/output error).  I expect this is a DERIVA configuration issue, but I didn't expect any user to be able to import the zip.

hategan commented:
Ok, so I rebased this (and the DMS) on auth_requests and used that scheme for the headers. It wasn't entirely smooth, since the DERIVA token does not have the resources_server pointing to the FQDN, so I had to override the default mechanism. 

@Xarthisius otherwise, auth_requests seems like a nice way of doing things and I approve of that general message. There were some inconsistencies in the verificators (some inherit from Verificator, some don't), but I guess that's fine in Python.

I also added settings wherever there were hardcoded URLs, as @craig-willis suggested. It is somewhat clear that, should this be adopted by the larger DERIVA community, there will be multiple URLs.

I did a manual test. At first I got an auth error inside the DMS. We have no way of reporting these to the user or offering the chance to take some pre-defined corrective action, but that's another story. It worked after I logged in and out of both deriva and Globus.

craig-willis reviewed (approved):
Basic import is working. After discussion with @Xarthisius, I recognize now that the issue with overwriting the existing dataset is actually bigger than just Deriva and will file a separate issue. From my understanding export/publish has yet to be implemented, so will leave those for future testing.
